libkbfs/block_ref_map.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2016 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libkbfs
+
+import "fmt"
+
+type blockContextMismatchError struct {
+	expected, actual BlockContext
+}
+
+func (e blockContextMismatchError) Error() string {
+	return fmt.Sprintf(
+		"Context mismatch: expected %s, got %s", e.expected, e.actual)
+}
+
+type blockRefEntry struct {
+	status  blockRefLocalStatus
+	context BlockContext
+	tag     interface{}
+}
+
+func (e blockRefEntry) checkContext(context BlockContext) error {
+	if e.context != context {
+		return blockContextMismatchError{e.context, context}
+	}
+	return nil
+}
+
+// blockRefMap is a map with additional checking methods.
+type blockRefMap map[BlockRefNonce]blockRefEntry
+
+func (refs blockRefMap) hasNonArchivedRef() bool {
+	for _, refEntry := range refs {
+		if refEntry.status == liveBlockRef {
+			return true
+		}
+	}
+	return false
+}
+
+func (refs blockRefMap) checkExists(context BlockContext) error {
+	refEntry, ok := refs[context.GetRefNonce()]
+	if !ok {
+		return blockNonExistentError{}
+	}
+
+	return refEntry.checkContext(context)
+}
+
+func (refs blockRefMap) getStatuses() map[BlockRefNonce]blockRefLocalStatus {
+	statuses := make(map[BlockRefNonce]blockRefLocalStatus)
+	for ref, refEntry := range refs {
+		statuses[ref] = refEntry.status
+	}
+	return statuses
+}
+
+func (refs blockRefMap) put(context BlockContext, status blockRefLocalStatus,
+	tag interface{}) error {
+	refNonce := context.GetRefNonce()
+	if refEntry, ok := refs[refNonce]; ok {
+		err := refEntry.checkContext(context)
+		if err != nil {
+			return err
+		}
+	}
+
+	refs[refNonce] = blockRefEntry{
+		status:  status,
+		context: context,
+		tag:     tag,
```

Comment:
This overrides the previous tag that might have been there, and I guess that's on purpose to reflect whatever the most recent operation's tag is.  Maybe we should call the `tag` field `mostRecentTag` so this isn't surprising?  And/or add some documentation to this file explaining `tag` semantics.